amundsen_gremlin/neptune_bulk_loader/gremlin_model_converter.py

```diff
@@ -0,0 +1,721 @@
+    make_table_uri
+)
+from amundsen_gremlin.utils.streams import chunk
+
+LOGGER = logging.getLogger(__name__)
+
+EXISTING_KEY = FrozenSet[Tuple[str, str]]
+EXISTING = NewType('EXISTING',
+                   Mapping[Union[VertexType, EdgeType], MutableMapping[EXISTING_KEY, Mapping[str, Any]]])
+ENTITIES = NewType('ENTITIES', Mapping[Union[VertexType, EdgeType], MutableMapping[str, Mapping[str, Any]]])
+
+
+def new_entities() -> ENTITIES:
+    return cast(ENTITIES, defaultdict(dict))
+
+
+def new_existing() -> EXISTING:
+    return cast(EXISTING, defaultdict(dict))
+
+
+def _get_existing_key_from_entity(_entity: Mapping[str, Any]) -> EXISTING_KEY:
+    """
+    Used for testing.
+    """
+    _entity = dict(_entity)
+    label: str = _entity.pop(MagicProperties.LABEL.value.name)
+    assert isinstance(label, str)  # appease the types
+    return _get_existing_key(_type=label, **_entity)
+
+
+def _get_existing_key(_type: Union[VertexType, EdgeType, VertexTypes, EdgeTypes, str], **_entity: Any) -> EXISTING_KEY:
+    """
+    Maybe this should be a part of EdgeType and VertexType.  But, this function certainly shouldn't be used
+    away from EXISTING (or testing)
+    """
+    if isinstance(_type, str):
+        vertex_type = VertexTypes.by_label().get(_type)
+        edge_type = EdgeTypes.by_label().get(_type)
+        assert bool(vertex_type) != bool(edge_type), \
+            f'expected exactly one of VertexTypes or EdgeTypes to match {_type}'
+        vertex_or_edge_type = vertex_type or edge_type
+        assert vertex_or_edge_type is not None  # appease mypy
+        _type = vertex_or_edge_type
+        assert isinstance(_type, (VertexTypes, EdgeTypes))
+    if isinstance(_type, (VertexTypes, EdgeTypes)):
+        _type = _type.value
+    assert isinstance(_type, (VertexType, EdgeType))
+    key_properties = _get_key_properties(_type)
+    # this eliding (the created and label fields) feels icky, but makes sense.
+    if isinstance(_type, EdgeType):
+        key_properties = key_properties.difference({WellKnownProperties.Created.value})
+    if isinstance(_type, VertexType):
+        key_properties = key_properties.difference({WellKnownProperties.TestShard.value})
+    key_properties = key_properties.difference({MagicProperties.LABEL.value})
+    assert WellKnownProperties.Created.value not in key_properties
+    assert MagicProperties.LABEL.value not in key_properties
+    return frozenset([(p.name, p.format(_entity.get(p.name))) for p in key_properties])
+
+
+@lru_cache(maxsize=len(list(VertexTypes)) + len(list(EdgeTypes)) + 100)
+def _get_key_properties(_type: Union[VertexType, EdgeType]) -> FrozenSet[Property]:
+    assert isinstance(_type, (EdgeType, VertexType))
+    return frozenset([_type.properties_as_map()[n] for n in _discover_parameters(_type.id_format)])
+
+
+def _discover_parameters(format_string: str) -> FrozenSet[str]:
+    """
+    use this to discover what the parameters to a format string
```

Comment:
similar comment as earlier: is this missing a word?
```suggestion
    use this to discover what the parameters to a format string are
```